content/posts/networkx/vf2pp/graph-iso-vf2pp/index.md

```diff
@@ -0,0 +1,247 @@
+```
+
+where "G2_nodes_of_degree" stores set of nodes for a given degree. The same is done with node labels.
+
+- **Extra shrinking of the candidate set for each node** by adding more checks in the candidate selection method and removing some from the feasibility checks. In simple terms, instead of checking a lot of conditions on a larger set of candidates, we check fewer conditions but on a more targeted and significantly smaller set of candidates.
+  For example, in this code:
+
+```python
+candidates = set(G2.nodes())
+for candidate in candidates:
+    if feasibility(u, candidate):
+        do_stuff()
+```
+
+we take a huge set of candidates, which results in poor performance due to maximizing calls of "feasibility", thus performing
+the feasibility checks in a very large set. Now compare that to the following alternative:
+
+```python
+candidates = [
+    n
+    for n in G2_nodes_of_degree[G1.degree[u]].intersection(
+        G2_nodes_of_label[G1_labels[u]]
+    )
+]
+for candidate in candidates:
+    if feasibility(u, candidate):
+        do_stuff()
+```
+
+Immediately we have drastically reduced the number of checks performed and calls to the function, as now we only apply them to nodes of the same degree and label as $u$. This is a simplification for demonstration purposes. In the actual implementation there are more checks and extra shrinking of the candidate set.
+
+## Demo
+
+Let's demonstrate our **VF2++** solver on a real graph. We are going to use the graph from the Graph Isomorphism wikipedia.
+
+<p float="center">
+  <img src="G1.png" width="200" height="200">
+  <img src="G2.png" width="395" height="250">
+</p>
+
+To use VF2++, we can do the following.
+
+```python
+import networkx as nx
+
+# edges of the graph
+edges = [
+    (1, 2),
+    (1, 5),
+    (1, 4),
+    (4, 8),
+    (2, 3),
+    (2, 6),
+    (3, 4),
+    (3, 7),
+    (5, 6),
+    (5, 8),
+    (6, 7),
+    (7, 8),
+]
+
+# map the nodes between the two graphs
+mapped_nodes = {1: "a", 2: "h", 3: "d", 4: "i", 5: "g", 6: "b", 7: "j", 8: "c"}
+
+# create the two networkx graphs
+G1 = nx.Graph(edges)
+G2 = nx.relabel_nodes(G1, mapped_nodes)
+
+# use the VF2++ without taking labels into consideration
+res = nx.vf2pp_is_isomorphic(G1, G2, node_label=None)
+# res: True
+res = nx.vf2pp_isomorphism(G1, G2, node_label=None)
+# res: {1: "a", 2: "h", 3: "d", 4: "i", 5: "g", 6: "b", 7: "j", 8: "c"}
+```
+
+Now if we wanted to take labels into consideration as well, we could do something like:
+
+```python
+import itertools
+
+labels = ["blue", "green", "yellow", "red", "brown", "black", "purple", "white"]
+# create the two networkx graphs
+G1 = nx.Graph(edges)
+G2 = nx.relabel_nodes(G1, mapped_nodes)
+
+# assign labels to the graphs (same label to mapped nodes)
+nx.set_node_attributes(G1, dict(zip(G1, itertools.cycle(labels))), "label")
+nx.set_node_attributes(
+    G2,
+    dict(zip([mapped_nodes[n] for n in G1.nodes()], itertools.cycle(labels))),
+    "label",
+)
```

Comment:
This is excellent - I really like the way things are presented first unlabelled then with labels. I have the same general comment as above though re: how the labels are applied: it's sort of "giving away the answer". Again, I'd suggest doing the more tedious manual labelling:

```suggestion
Now if we wanted to take labels into consideration as well, we could do something like:

```python
G_nattrs = {
    "a": "blue",
    "g": "green",
    "b": "pink",
    "h": "red",
    "c": "yellow",
    "i": "orange",
    "d": "cyan",
    "j": "purple",
}
nx.set_node_attributes(G, G_nattrs, name="color")

H_nattrs = {
    1: "blue",
    2: "red",
    3: "cyan",
    4: "orange",
    5: "green",
    6: "pink",
    7: "purple",
    8: "yellow",
}
nx.set_node_attributes(H, H_nattrs, name="color")
```